xarray/testing/strategies.py

```diff
@@ -0,0 +1,343 @@
+from collections.abc import Hashable, Mapping, Sequence
+from typing import Any, Protocol, Union
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+import numpy as np
+
+import xarray as xr
+from xarray.core.types import T_DuckArray
+
+__all__ = [
+    "numeric_dtypes",
+    "names",
+    "dimension_names",
+    "dimension_sizes",
+    "attrs",
+    "variables",
+]
+
+
+class ArrayStrategyFn(Protocol):
+    def __call__(
+        self, *, shape: tuple[int, ...] = None, dtype: np.dtype = None, **kwargs
+    ) -> st.SearchStrategy[T_DuckArray]:
+        ...
+
+
+# required to exclude weirder dtypes e.g. unicode, byte_string, array, or nested dtypes.
+def numeric_dtypes() -> st.SearchStrategy[np.dtype]:
+    """
+    Generates only those numpy dtypes which xarray can handle.
+
+    Requires the hypothesis package to be installed.
+    """
+
+    return (
+        npst.integer_dtypes()
+        | npst.unsigned_integer_dtypes()
+        | npst.floating_dtypes()
+        | npst.complex_number_dtypes()
+    )
+
+
+def np_arrays(
+    *,
+    shape: Union[
+        tuple[int, ...], st.SearchStrategy[tuple[int, ...]]
+    ] = npst.array_shapes(max_side=4),
+    dtype: Union[np.dtype, st.SearchStrategy[np.dtype]] = numeric_dtypes(),
+) -> st.SearchStrategy[np.ndarray]:
+    """
+    Generates arbitrary numpy arrays with xarray-compatible dtypes.
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    shape
+    dtype
+        Default is to use any of the numeric_dtypes defined for xarray.
+    """
+
+    return npst.arrays(dtype=dtype, shape=shape)
+
+
+# TODO Generalize to all valid unicode characters once formatting bugs in xarray's reprs are fixed + docs can handle it.
+_readable_characters = st.characters(
+    categories=["L", "N"], max_codepoint=0x017F
+)  # only use characters within the "Latin Extended-A" subset of unicode
```

Comment:
We also may be able to relax this at some point. Right now if you remove these restrictions you get some lovecraftian-level reprs!

```python
In [5]: variables().example()
Out[5]: 
<xarray.Variable (é|¹ã: 2, 8񱮰·: 4, 󾌦eFð: 1)>
array([[[2.0000100e+00+0.0000000e+00j],
        [0.0000000e+00-0.0000000e+00j],
        [          inf+0.0000000e+00j],
        [          nan          +infj]],

       [[3.4028235e+38          +nanj],
        [5.1881745e+16-1.9000000e+00j],
        [2.2204460e-16-0.0000000e+00j],
        [0.0000000e+00+6.1035156e-05j]]], dtype=complex64)
Attributes: (12/13)
    oÜ霎:     True
    󒡊=        False
    .􁆭:       ÍÂ
    :         [0.]
    :      [[-0.0000000e+00+0.j   6.1035156e-05-1.9j]]
    ...       ...
    U«🧬񙾢#:    None
    ?ìÅ
:    True
    %ñ:      [46289]
    è𬭏â:     
    􎿑Ôq¿�:    None
    ÿ[¹򺟚:     False
```